src/lib/ui/Menu/Menu.tsx

```diff
@@ -0,0 +1,179 @@
+import classNames from 'classnames'
+import {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useLayoutEffect,
+  useState,
+} from 'react'
+import { createPortal } from 'react-dom'
+
+export interface MenuProps {
+  children: JSX.Element | JSX.Element[]
+  verticalOffset?: number
+  horizontalOffset?: number
+  edgeOffset?: number
+  button: (props: {
+    open: (event: React.MouseEvent<HTMLElement>) => void
+  }) => React.ReactElement
+  BackgroundProps?: Partial<{
+    className?: string
+  }>
+}
+
+interface MenuContextProps {
+  close: () => void
+}
+
+const MenuContext = createContext<MenuContextProps | undefined>(undefined)
+
+const Menu = ({
+  verticalOffset = 5,
+  horizontalOffset = 0,
+  edgeOffset = 5,
+  children,
+  button,
+  BackgroundProps,
+}: MenuProps) => {
+  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null)
+  const [documentEl, setDocumentEl] = useState<null | Element>(null)
+  const [contentEl, setContentEl] = useState<HTMLDivElement | null>(null)
+  const [top, setTop] = useState(0)
+  const [left, setLeft] = useState(0)
+
+  useEffect(() => {
```

Comment:
Did we consider using `useLayoutEffect`? Might be appropriate? https://kentcdodds.com/blog/useeffect-vs-uselayouteffect

Reply:
Usually only do `useLayoutEffect` when something seems off due to render timings. Doesn't hurt to switch to it though.

Reply:
Ok, I see you use it below, so maybe this case is correct.